### ISSUE TITLE
Inconsistent result for Spark Round function

### ISSUE THREAD

**Issue by rui-mo** [facebookincubator/velox#10929]
### Bug description

We noticed that the result was inconsistent with Spark for `round(0.575, 2)`. The desired result is 0.58, but since double cannot accurately represent 0.575, which is 0.57499999999999999, we obtained 0.57. https://github.com/facebookincubator/velox/pull/5992 was proposed to fix this case, which uses `std::nextafter` to enlarge the input slightly, yielding a result of 0.58. This change was not accepted by Velox since receiving double means we can not recover the original value.

We introduced this function internally to Gluten to handle the majority of cases; but for a case like `round (0.5549999999999999, 2)`, the result will become 0.56 rather than 0.55. This prompts us to reconsider this issue. One idea is to map the Spark double type to a higher-precision C++ type, such as `boost::multiprecision::cpp_dec_float_50`, but this will take a lot of changes in Velox.

The following table lists the expected and actual results from the Velox and Gluten rounds. Both cases can be reproduced with https://github.com/facebookincubator/velox/commit/8408d33121eb06695330ec5c3c7b75711c4453a8.

| Input  | Round to scale | Expected | Round | Round with nextafter | 
| ------------- | ------------- | ------------- | ------------- | ------------- |
| 0.575  | 2  | 0.58 | 0.57 | 0.58 |
| 0.5549999999999999  | 2  | 0.55 | 0.56 | 0.56 |
| 0.499999999999994  | 0  | 0 | 1 | 0 |

\* Round: refers to the [round](https://github.com/facebookincubator/velox/blob/94dcf028c12a892b753946563edcc4589f69fe79/velox/functions/prestosql/Arithmetic.h#L219) implementation under presosql in Velox.
\* Round with nextafter: refers to the Gluten proposed [round](https://github.com/rui-mo/velox/blob/fc5889f6e276109b40645bdcddde61ca71f4c7fd/velox/functions/sparksql/Arithmetic.h#L32) implementation with nextafter used.

### System information

Velox System Info v0.0.2
Commit: fc5889f6e276109b40645bdcddde61ca71f4c7fd
CMake Version: 3.28.3
System: Linux-5.4.0-189-generic
Arch: x86_64
C++ Compiler: /usr/bin/c++
C++ Compiler Version: 11.1.0
C Compiler: /usr/bin/cc
C Compiler Version: 11.1.0
CMake Prefix Path: /usr/local;/usr;/;/usr/local/lib/python3.8/dist-packages/cmake/data;/usr/local;/usr/X11R6;/usr/pkg;/opt

**Comment by rui-mo** [comment]
@majetideepak Opened this issue for the round result issue we met and added reproducible tests. Could you spare some time to take a look? Thank you.

**Comment by ViggoC** [comment]
@rui-mo So, adding afternext can't completely solve this issue, right? And when you say "map the Spark double type to a higher-precision C++ type", you mean just in some functions like round or every expressions?

**Comment by rui-mo** [comment]
@ViggoC Yes, 'nextafter' cannot solve the high-precision cases. I assume this solution requires a complete replacement for the double type, because once the precision gets lost during computing, the original precision cannot be recovered. Alternatively, we might recommend customers use the decimal type if they have a requirement for high precision.

**Comment by ViggoC** [comment]
@rui-mo You mean cast all floating-point numbers to a high-precision type, and do the expression calculation base on the high-precision and cast then back to the origin precision before return result. But no matter how high the precision of a floating-point number is, it cannot accurately represent a decimal value. The high-precision type should be Decimal is no precision loss is accepted. So that the difference between the two solutions at this point is that the user specifies through a configuration whether to use decimal for internal expression calculations or explicitly convert all floating-point number fields to decimal.

**Comment by ViggoC** [comment]
@rui-mo In our internal Velox repository, we have aligned Spark's behavior by converting it to BigDecimal. If we only consider the round function, I think this is enough.

**Comment by rui-mo** [comment]
@ViggoC Thank you so much for sharing this information. May I ask if you are following

**Comment by rui-mo** [comment]
@ViggoC Thank you so much for sharing this information. May I ask if you are following the below conversion in BigDecimal or using a similar conversion as the cast implementation in Velox? Just wondering if it is too heavy to implement as the BigDecimal one.

https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/math/BigDecimal.java#L991-L1007

https://github.com/facebookincubator/velox/blob/c04652462250469d759396aac9d68d441ac1eb2a/velox/expression/CastExpr-inl.h#L426

**Comment by ViggoC** [comment]
@rui-mo We implemented it by converting to string and then parsing, not sure if this is the best practice. You can do a performance comparison, it should not be complicated.

**Comment by rui-mo** [comment]
Hi @majetideepak, to cover more cases, do you think it makes sense to convert floating-point numbers as decimal before round? Similar to Spark, more cases could be covered, but not all of them because the floating-point input for the to-decimal conversion is not precise.

**Comment by majetideepak** [comment]
@rui-mo As @ViggoC mentioned, we need to convert the string-literal to a decimal value so as to not lose any precision.
Regarding `round(cast(0.575 as double), 2)`, the results can vary between C++ and Java since the double semantics vary. Decimal is expected to be the same.
However, in practice, for Presto, a string-literal is converted to a decimal value at the coordinator before being passed to worker(Velox).
How does this work with Spark? Where does the string to double conversion happen?

**Comment by rui-mo** [comment]
> How does this work with Spark? Where does the string to float/double conversion happen?

@majetideepak Thanks for your answering. Spark converts float/double as BigDecimal and uses its 'setScale' function for the rounding. I also notice Presto uses 'roundToLong' function and converts the input as BigDecimal if the value is out of the range of long.

Presto round: https://github.com/prestodb/presto/blob/08cf9bbceece63703a51cef6222e6b0a7fdd4f1e/presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java#L1099-L1117

Spark round: https://github.com/apache/spark/blob/c374c00432bd28b98b7623481b7985aa7bc7624a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala#L1587-L1600

In Velox to convert float/double as decimal, @ViggoC shared their practice to convert float/double as string first then as decimal. We could also convert float/double as decimal directly like Spark does.

**Comment by majetideepak** [comment]
> their practice to convert float/double as string first then as decimal.

I don't think this will help resolve the precision issue. 

> We could also convert float/double as decimal directly like Spark does.

You mean float/double literal value correct?

**Comment by rui-mo** [comment]
> You mean float/double literal value correct?

The Spark round implementation converts float/double inputs as decimal firstly, then round them as the needed scale. This conversion is for all floating-point inputs and not limited to literals.
